tests/test_collector_manager.py

```diff
@@ -0,0 +1,467 @@
+    for sink in sinks:
+        value = sink.args
+        assert value
+
+
+@pytest.mark.asyncio
+async def test_collector_manager_collect_filesink(
+    mock_entity: MockCollectorEntityFunc,
+    mock_redis,
+    monkeypatch,
+    mock_mem,
+    mock_con,
+) -> None:
+    """Ensure that all collectors are executed and some metric is retrieved
+    and the FileSink is written to as expected"""
+    entity1 = mock_entity(port=1234, name="entity1", telemetry_on=True)
+    entity2 = mock_entity(port=2345, name="entity2", telemetry_on=True)
+
+    sinks = [
+        FileSink(entity1.status_dir + "/1_con.csv"),
+        FileSink(entity1.status_dir + "/1_mem.csv"),
+        FileSink(entity2.status_dir + "/2_mem.csv"),
+    ]
+    con_col1 = DBConnectionCollector(entity1, sinks[0])
+    mem_col1 = DBMemoryCollector(entity1, sinks[1])
+    mem_col2 = DBMemoryCollector(entity2, sinks[2])
+
+    manager = CollectorManager()
+    manager.add_all([con_col1, mem_col1, mem_col2])
+
+    # Execute collection
+    with monkeypatch.context() as ctx:
+        ctx.setattr(
+            redisa,
+            "Redis",
+            mock_redis(client_stats=mock_con(1, 10), mem_stats=mock_mem(1, 10)),
+        )
+        await manager.collect()
+
+    # verify each collector retrieved some metric & sent it to the sink
+    for sink in sinks:
+        save_to = sink.path
+        assert save_to.exists()
+        if "con" in str(save_to):
+            assert "127.0.0." in save_to.read_text()
+        else:
+            # look for something multiplied by 1000
+            assert "000" in save_to.read_text()
+
+
+@pytest.mark.asyncio
+async def test_collector_manager_collect_integration(
+    test_dir: str, mock_entity: MockCollectorEntityFunc, local_db, mock_sink
+) -> None:
+    """Ensure that all collectors are executed and some metric is retrieved"""
+    entity1 = mock_entity(port=local_db.ports[0], name="e1")
+    entity2 = mock_entity(port=local_db.ports[0], name="e2")
+
+    # todo: consider a MockSink so i don't have to save the last value in the collector
+    sinks = [mock_sink(), mock_sink(), mock_sink()]
+    con_col1 = DBConnectionCollector(entity1, sinks[0])
+    mem_col1 = DBMemoryCollector(entity1, sinks[1])
+    mem_col2 = DBMemoryCollector(entity2, sinks[2])
+
+    manager = CollectorManager()
+    manager.add_all([con_col1, mem_col1, mem_col2])
+
+    # Execute collection
+    await manager.collect()
+
+    # verify each collector retrieved some metric & sent it to the sink
+    for sink in sinks:
+        value = sink.args
+        assert value
+
+
+@pytest.mark.parametrize(
+    "timeout_at,delay_for,expect_fail",
+    [
+        pytest.param(1000, 5000, True, id="1s timeout"),
+        pytest.param(2000, 5000, True, id="2s timeout"),
+        pytest.param(3000, 5000, True, id="3s timeout"),
+        pytest.param(4000, 5000, True, id="4s timeout"),
+        pytest.param(2000, 1000, False, id="under timeout"),
```

Comment:
I cannot explain _why_, but I am consistently getting "un-ended task" warnings on the `3s timeout` and `under timeout`. It shouldn't be _that_ big a deal as it looks like it confined to the test suite (or at least I couldn't reproduce with any of the experiments I ran), but it might be worth looking into / documenting in a ticket as "known issue".

I'll keep looking into this in the meantime to see if I can't offer some actually useful help, lol!

Reply:
The first think I might try is adding an `await asyncio.gather(*manger._tasks)` at the end of the test fn just to make sure all of the coroutines have a chance to properly be canceled, but that is honestly just a guess, hahaha

Reply:
I forgot that `asyncio.wait` doesn't cancel tasks and `asyncio.wait_for` does. I added a loop to cancel the pending tasks after the timeout period and the errors are gone.